### ISSUE TITLE
Refactor individual API

### ISSUE THREAD

**Issue by giovannic** [mrc-ide/individual#46]
The API adds a lot of boilerplate to adding new variables (like categorical, vector-based, multi-category). The API is also complicated to mock.

So lets refactor it out. The high level changes will include:

 * No more `SimAPI` or api objects. Clients will instead make their variables and pass them into their processes.
 * No more `State` object. This is just another variable, it's not special. There could be several state-like variables.

So the basic SIR example goes from:

```r
population <- 4
S <- State$new('S', population)
I <- State$new('I', 0)
R <- State$new('R', 0)
human <- Individual$new('human', list(S, I, R))

transition <- function(from, to, rate) {
  return(function(api) {
    from_state <- api$get_state(human, from)
    api$queue_state_update(
      human,
      to,
      from_state[runif(length(from_state), 0, 1) < rate]
    )
  })
}

processes <- list(
  transition(S, I, .2),
  transition(I, R, .1),
  transition(R, S, .05),
  state_count_renderer_process(individual$name, c('S', 'I', 'R'))
)

simulate(human, processes, 5)
```

To...

```r
population <- 4
timesteps <- 5
state <- CategoricalVariable$new(c('S', 'I', 'R'), rep('S', population))
renderer <- Render$new(timesteps)

transition <- function(from, to, rate) {
  return(function(api) {
    from_state <- state$get_index_of(from)
    state$queue_update(
      to,
      from_state$sample(rate)
    )
  })
}

processes <- list(
  transition('S', 'I', .2),
  transition('I', 'R', .1),
  transition('R', 'S', .05),
  categorical_count_renderer_process(renderer, state, c('S', 'I', 'R'))
)

simulation_loop(variables=list(state), processes=processes, timesteps=timesteps)
renderer$to_dataframe()
```

Going forward, this decouples the variables and allows us to improve them and create new ones in isolation

**Comment by giovannic** [comment]
This is in progress on [this branch](https://github.com/mrc-ide/individual/tree/feat/variable_refactor)

**Comment by giovannic** [comment]
Closes #42 #37 #30 #32 #29 #26
